### ISSUE TITLE
Isokinetic momentum update loses up to 19 nats near antiparallel momentum; use the non-negative form

### ISSUE THREAD

**Issue by junpenglao** [blackjax-devs/blackjax#1038]
Follow-up from the adversarial review of #1036. That PR fixes the *downstream* consequence (#1035); this is the source.

## The defect

`blackjax/mcmc/integrators.py:459,470-473` computes the isokinetic momentum update's log argument as

```python
A = 1 + momentum_proj + (1 - momentum_proj) * zeta**2
```

`momentum_proj` is a dot product of two unit vectors, so `|momentum_proj| <= 1` mathematically — but in float32 it can round marginally below `-1`, making `A` negative and `jnp.log(A)` a NaN.

Instrumenting all **204,500** momentum updates of a full 1000-iteration, 100-chain LAPS burn-in: each failing run contains **exactly one** call with `A <= 0`, always at

```
momentum_proj = -1.0000001192092896   # exactly -(1 + 2**-23), one float32 ulp below -1
```

With `p = -(1+u)`, `u = 2**-23`: `A = -u + (2+u)*zeta**2`, so `A < 0` iff `zeta < sqrt(u/2) = 2.44e-4`. That condition separates failing from healthy runs exactly — a healthy key had two `p < -1` events whose `zeta` was too large to trigger it. `p < -1` is necessary but not sufficient.

## Do not "fix" this by clamping

The obvious repair, `momentum_proj = jnp.clip(momentum_proj, -1.0, 1.0)`, does remove the NaN — but it converts a loud failure into a quiet one. Scored against a float64 oracle over 200,000 float32 cases with the momentum within `1e-8 .. 1` rad of antiparallel to the gradient:

| form | `A < 0` | `log(A)` non-finite | median abs err in `log A` | p99 | max |
|---|---|---|---|---|---|
| current `1 + p + (1-p)z²` | 623 | 623 | 4.45e-05 | **14.3** | 19.4 |
| clamp `p <- clip(p,-1,1)` | 0 | 0 | 4.42e-05 | **14.3** | 19.4 |
| `A = ½(‖m+ĝ‖² + z²‖m−ĝ‖²)` | 0 | 0 | **2.53e-07** | **0.123** | 2.18 |

The clamp leaves up to **19 nats** of error in the energy accounting, which flows into `EEVPD` and the step-size adaptation with nothing to catch it. That is strictly worse than the status quo, where the NaN is at least visible.

## Proposed fix

Use the algebraically identical form, via `‖m ± ĝ‖² = 2 ± 2p`:

```python
A = 0.5 * (sq_norm(momentum + grad_unit) + zeta**2 * sq_norm(momentum - grad_unit))
```

Non-negative by construction (so no clamp is needed), cancellation-free as `p -> -1`, and it cuts the p99 error from 14.3 nats to 0.12. Cost: two length-`d` squared norms replacing one length-`d` dot product — same order as what is already there.

## Keep the #1036 guards regardless

There is a residual route this does not close: `zeta` underflows to exactly `0.0` on ~1500 of 204,500 calls per run. If `p == -1.0` exactly co-occurs with `zeta == 0`, then `A == 0` and `log(0) = -inf`, not NaN. That did not occur in the four instrumented runs but is reachable — and it is the concrete justification for #1036 choosing `isfinite` over `isnan` in the NaN gate.

## Verification

Any candidate should be scored against a float64 oracle on near-antiparallel cases as above, and checked bit-identical on runs where `p > -1` throughout (the change must be a no-op on the ~204,499 calls that are already well-conditioned).

— 🤖 Blackjax-devs AI TL

**Comment by junpenglao** [comment]
**Correction: the form proposed above is wrong at zero gradient. Do not implement it as written.**

`_normalized_flatten_array` (`integrators.py:373`) returns `x` itself, not a unit vector, when the gradient norm is `<= 1e-13`. So at a zero gradient `grad_unit` is the zero vector, and the identity `||m ± g||² = 2 ± 2p` that the proposal rests on fails by O(1). There, `zeta = 1` and:

| form | `A` at zero gradient | kinetic-energy error |
|---|---|---|
| current | 2 | — (correct) |
| proposed above | 1 | `-(d-1)·log 2` — **-692 nats at d=1000** |

This fires whenever a trajectory starts at a stationary point — a `jnp.zeros` init on a centred target, or a MAP init. Measured effect with adjusted LRD warmup at D=6: pilot n_eff collapses from 213 / 296 / 104 to 4.5 / 47 / 42 (seeds 50/51/52), and recovers exactly with a `1e-3 * ones` init. Five tests in `test_mclmc_lrd_adaptation.py` fail under it — all from `jnp.zeros` on a zero-mean Gaussian, so they are real catches, not drift.

A second issue: dropping to one squared norm is not safe either. `isokinetic_omelyan` has momentum coefficient `b3 = -0.266`, which drives `zeta > 1`, and `laps.py` uses omelyan by default for d > 200. In that regime `(1 - p)·zeta²` is the term that cancels.

**Recommended instead — a two-sided form**, with `S± = 0.5 * ||m ± g||²`:

```python
A = jnp.where(zeta**2 <= 1,
              S_plus * (1 - zeta**2) + 2 * zeta**2,
              S_minus * (zeta**2 - 1) + 2)
```

Each branch is exact for unit vectors, non-negative by construction, and computes whichever of `(1+p)` / `(1-p)` is cancelling as a squared norm. At zero gradient it gives `A = 2`, matching the current code. Measured: zero `A < 0` cases in every regime tested (including omelyan's `zeta >= 3000` corner, where current code has 206/20k), accuracy tied with the best alternative (p99 error in `log A` 2.1e-4 vs current 8.7 nats), and all 19 affected test files pass (977, same as baseline). Cost is two extra squared norms — the same as the original proposal.

Two corrections to the verification section above:

- **Bit-identity cannot be met.** Only 32–41% of real calls are bit-identical under any stable form, since the change perturbs well-conditioned calls too. Use a tolerance criterion. No golden values are affected; the tight asserts compare the same formula on both sides.
- **The norm-defect effect is benign at O(eps).** Measured `||m||² - 1` over a real burn-in: median 6e-8, max 4.8e-7. To first order the discrepancy is proportional to `(1+p)`, so it vanishes exactly where the cancellation happens — folding the real norm in is an improvement, not a semantic change. Note an oracle that does not renormalise is itself negative on 13.6% of near-antiparallel cases, so it cannot serve as the reference.

Not yet measured: runtime cost (expected negligible under fusion), and how often a real d > 200 LAPS run reaches `zeta > 1`.

— 🤖 Blackjax-devs AI TL
